src/clipboard.c

```diff
@@ -0,0 +1,538 @@
+typedef struct _ClipboardClass ClipboardClass;
+
+struct _Clipboard
+{
+  XdpDbusClipboardSkeleton parent_instance;
+};
+
+struct _ClipboardClass
+{
+  XdpDbusClipboardSkeletonClass parent_class;
+};
+
+static XdpDbusImplClipboard *impl;
+static Clipboard *clipboard;
+
+GType clipboard_get_type (void) G_GNUC_CONST;
+static void clipboard_iface_init (XdpDbusClipboardIface *iface);
+
+G_DEFINE_TYPE_WITH_CODE (Clipboard,
+                         clipboard,
+                         XDP_DBUS_TYPE_CLIPBOARD_SKELETON,
+                         G_IMPLEMENT_INTERFACE (XDP_DBUS_TYPE_CLIPBOARD,
+                                                clipboard_iface_init))
+
+static XdpOptionKey clipboard_set_selection_options[] = {
+  { "mime_types", G_VARIANT_TYPE_STRING_ARRAY, NULL },
+};
+
+static gboolean
+handle_request_clipboard (XdpDbusClipboard *object,
+                          GDBusMethodInvocation *invocation,
+                          const char *arg_session_handle,
+                          GVariant *arg_options)
+{
+  Call *call = call_from_invocation (invocation);
+  Session *session;
+  RemoteDesktopSession *remote_desktop_session;
+
+  session = acquire_session_from_call (arg_session_handle, call);
+  if (!session)
+    {
+      g_dbus_method_invocation_return_error (invocation,
+                                             G_DBUS_ERROR,
+                                             G_DBUS_ERROR_ACCESS_DENIED,
+                                             "Invalid session");
+      return G_DBUS_METHOD_INVOCATION_HANDLED;
+    }
+
+  SESSION_AUTOLOCK_UNREF (session);
+
+  if (!is_remote_desktop_session (session))
+    {
+      g_dbus_method_invocation_return_error (invocation,
+                                             G_DBUS_ERROR,
+                                             G_DBUS_ERROR_ACCESS_DENIED,
+                                             "Invalid session type");
+      return G_DBUS_METHOD_INVOCATION_HANDLED;
+    }
+  remote_desktop_session = (RemoteDesktopSession *)session;
+
+  if (!remote_desktop_session_can_request_clipboard (remote_desktop_session))
+    {
+      g_dbus_method_invocation_return_error (
+          invocation, G_DBUS_ERROR, G_DBUS_ERROR_FAILED, "Invalid state");
+      return G_DBUS_METHOD_INVOCATION_HANDLED;
+    }
+
+  xdp_dbus_impl_clipboard_call_request_clipboard (
+      impl, session->id, arg_options, NULL, NULL, NULL);
+
+  xdp_dbus_clipboard_complete_request_clipboard (object, invocation);
+  remote_desktop_session_clipboard_requested (remote_desktop_session);
+
+  return G_DBUS_METHOD_INVOCATION_HANDLED;
+}
+
+static gboolean
+handle_set_selection (XdpDbusClipboard *object,
+                      GDBusMethodInvocation *invocation,
+                      const char *arg_session_handle,
+                      GVariant *arg_options)
+{
+  Call *call = call_from_invocation (invocation);
+  Session *session;
+  GVariantBuilder options_builder;
+  GVariant *options;
+  g_autoptr (GError) error = NULL;
```

Comment:
The coding style for `g_autoptr` in xdg-desktop-portal is to not have a space before `(`.